### PR TITLE
Use non-recursive DFS when visit binary expression

### PR DESCRIPTION
**Pull request by Cosifne**
This recursion is crashing devenv when using [LongClass.txt](https://github.com/dotnet/roslyn/files/9312912/LongClass.txt) created by @dibarbet . 

It is found when we investigate this feedback ticket https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1582837
However, this is not a regression, so I am fine if we decide not to take this.

### CODE REVIEW

File: src/Features/Core/Portable/Wrapping/BinaryExpression/AbstractBinaryExpressionWrapper.cs [hunk 1]
```diff
@@ -7,16 +7,17 @@
 using System.Threading.Tasks;
 using Microsoft.CodeAnalysis.LanguageService;
 using Microsoft.CodeAnalysis.PooledObjects;
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis.Indentation;
+using Microsoft.CodeAnalysis.Precedence;
+using Roslyn.Utilities;
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Wrapping/BinaryExpression/AbstractBinaryExpressionWrapper.cs:13]: Sort

File: src/Features/Core/Portable/Wrapping/BinaryExpression/AbstractBinaryExpressionWrapper.cs [hunk 2]
```diff
@@ -103,19 +104,52 @@ internal abstract partial class AbstractBinaryExpressionWrapper<TBinaryExpressio
...
-            {
-                _syntaxFacts.GetPartsOfBinaryExpression(
-                    expr, out var left, out var opToken, out var right);
-                AddExpressionsAndOperators(precedence, left, result);
-                result.Add(opToken);
-                AddExpressionsAndOperators(precedence, right, result);
-            }
-            else
+            // In-order traverse which visit the left child -> operator in the binary expression -> right child
+            using var pooledStack = SharedPools.Default<Stack<(SyntaxNode node, bool isLeafNode)>>().GetPooledObject();
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Wrapping/BinaryExpression/AbstractBinaryExpressionWrapper.cs:108]: i've been trying to wrap my head around this, and it seems off...

to me, isn't this just a simple use of a stack like this:

1. add the initial node to the stack (which is a stack of SyntaxNodeOrToken)
2. loop as long as the stack is not empty
3. pop the top item off of the stack.  if it's a token, just add to the result
4. if it's a node then see if is a bin-expr of the same precedence.  
5. if not, add the node to the result
6. if so, break the node into `left, op, right`.  push onto the stack in reverse order.
7. repeat loop

Wouldn't this work?

Comment (Cosifne) [reply src/Features/Core/Portable/Wrapping/BinaryExpression/AbstractBinaryExpressionWrapper.cs:108]: It feels like they are the same algorithm, except my version would not keep the `right` node in the stack

Comment (Cosifne) [reply src/Features/Core/Portable/Wrapping/BinaryExpression/AbstractBinaryExpressionWrapper.cs:108]: Example:
```
     A
    /  \
  B    C
```
In my version the stack state would be:
`A`
`B A` (B visited) 
`A`  (A visited)
`C` (C visited)
And your version would be:
`B` `A` `C` (B visited)
`A` `C` (A visited)
`C` (C visited)

If

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/Wrapping/BinaryExpression/AbstractBinaryExpressionWrapper.cs:108]: rigth.  imo, my algorithm is simpler.  there's no need for multiple pieces of tracking state (e.g. 'currentNode' *and* the stack).  ANd my algorithm doesn't need a 'isLeafNode' boolean).

Comment (Cosifne) [reply src/Features/Core/Portable/Wrapping/BinaryExpression/AbstractBinaryExpressionWrapper.cs]: As long as we don't use the recursion it's fine to me.
I have updated the version.

**Review by CyrusNajmabadi — APPROVED** [review]

**Pull request merged**